### PR TITLE
feat: add Doxygen

### PR DESCRIPTION
**Pull request by amaanq**

### CODE REVIEW

File: queries/c/injections.scm [hunk 1]
```diff
@@ -8,6 +8,10 @@
   (#match? @injection.content "/\*!([a-zA-Z]+:)?re2c")
   (#set! injection.language "re2c"))
 
+((comment) @injection.content
+  (#match? @injection.content "/\*(!|<|\*)[^\w]")
```
Comment (lucario387) [inline queries/c/injections.scm:12]: Shouldn't all of these patterns be `^pattern`

Comment (amaanq) [reply queries/c/injections.scm:12]: /** /*! and /*< signify a potential doxygen blob

I added a not word after to avoid double injecting w/ re2c

Comment (lucario387) [inline queries/c/injections.scm:12]: Also, shouldn't all of these be escaped twice i.e. `\\w`

**Review by lucario387 — APPROVED** [review]

**Comment by amaanq** [comment]
Seems like it just needs indent.auto to play nicely

**Review by ObserverOfTime — COMMENTED** [review]
Can you also highlight [links and references](https://www.doxygen.nl/manual/autolink.html)?

**Comment by amaanq** [comment]
Done, though w/o self.parent to inject the function params the entire blob will be highlighted as function

**Comment by ObserverOfTime** [comment]
nvim crashes when I open a cpp file with the doxygen parser installed. :thinking:

**Comment by amaanq** [comment]
Hm, fuzzing shows nothing - can you provide a reproducer and potentially a backtrace? Maybe it's certain injections inside doxygen

**Comment by ObserverOfTime** [comment]
Can you reproduce with this?
```cpp
#ifdef NDEBUG

#endif  // NDEBUG
```

**Comment by amaanq** [comment]
I cannot

**Comment by ObserverOfTime** [comment]
Then it's probably one of my plugins. :thinking:

**Comment by amaanq** [comment]
does it crash just *opening* a file w/ that text or *after* you make some edits?

(leading to me wanting to add a fuzzer for edits feature upstream even more so if the latter..)

**Comment by ObserverOfTime** [comment]
> does it crash just opening a file w/ that text or after you make some edits?

Both after opening a file w/ that text and after writing one (I think it's the comment specifically).

**Comment by ObserverOfTime** [comment]
It crashes even with `--clean -u scripts/minimal_init.lua` so it's not a plugin that causes it.
That leaves either another parser or some other config file. :thinking:

**Comment by ObserverOfTime** [comment]
Interesting. It only crashes when I write a word character in the comment.

**Comment by amaanq** [comment]
maybe it's the comment parser if you use it? a gdb/lldb backtrace could be really insightful here

**Comment by ObserverOfTime** [comment]
Works after escaping the asterisks twice. Wasn't my issue after all. :)

**Comment by amaanq** [comment]
In the match? It's still really weird behavior though

**Comment by amaanq** [comment]
Hmm, is it good now @ObserverOfTime?

**Comment by ObserverOfTime** [comment]
Unfortunately not… I'm encountering another crash.

**Comment by ObserverOfTime** [comment]
Write this in a cpp file and press enter anywhere after the `!`. 
```cpp
/*! \brief Help */
```

**Comment by amaanq** [comment]
doesn't crash for me

**Comment by ObserverOfTime** [comment]
This is very weird.

**Comment by ObserverOfTime** [comment]
I'll leave it for tomorrow…

**Comment by ObserverOfTime** [comment]
Here's the backtrace. I don't know if it helps.
image

**Comment by amaanq** [comment]
Yeah that's very helpful, and wow, ironic how someone else reported that same bug and I fixed it as well

https://github.com/tree-sitter/tree-sitter/issues/2563

https://github.com/tree-sitter/tree-sitter/pull/2566

If possible - can you set the ts dep for neovim to my PR's branch and try it out then?

**Comment by amaanq** [comment]
I see a potential spot where this can occur in the scanner too, I've push an update, can you try that now?

**Comment by ObserverOfTime** [comment]
It's fixed after `!` but pressing enter before the closing `*/` still causes it to crash.

Also, you can change the matches to this (since re2c only matches letters):
```scheme
(#lua-match? @injection.content "/[*][!<*][^a-zA-Z]")
```

**Comment by amaanq** [comment]
> It's fixed after `!` but pressing enter before the closing `*/` still causes it to crash.

Found another spot, but did you also build neovim w/ my ts fork?

**Comment by ObserverOfTime** [comment]
> Found another spot, but did you also build neovim w/ my ts fork? (It'd be nice to confirm if that fixes it)

That does seem to fix it.

**Comment by amaanq** [comment]
The latest doxygen fix (2nd one) or my ts pr?

**Comment by ObserverOfTime** [comment]
Both.

**Comment by amaanq** [comment]
awesome, thanks

**Comment by ObserverOfTime** [comment]
The `tag_name` is sometimes not highlighted. Might need a higher priority.

```scheme
((tag_name) @keyword
 (#set! "priority" 101))
```

**Comment by amaanq** [comment]
I notice that too - adding any item below it (like a function or something) fixes it. I'd want to say it's not a bug w/ priorities

**Comment by ObserverOfTime** [comment]
Maybe, but the priority fixes it for me.

**Comment by amaanq** [comment]
Just tag name, or all of them?

**Comment by ObserverOfTime** [comment]
The others seem fine.

**Review by ObserverOfTime — APPROVED** [review]
Finally…

**Comment by amaanq** [comment]
the joy of debugging tree-sitter

**Pull request merged**